### ISSUE TITLE
Request for wrapping std::vector

### ISSUE THREAD

**Issue by lambage** [OldFisher/lua-api-pp#6]
Is it possible to wrap std::vector types (ideally just float, string, int etc...) to automatically convert to tables?  This library is great but I feel this is something it's possibly missing.  I'm just looking to be able to do something like:

std::vector<int> read_list(size_t count);
void write_list(std::vector<int> list);

...

c.global["read_list"] = read_list;
c.global["write_list"] = write_list;

I'll look into seeing if it's something I can maybe try developing myself but was just wondering if anyone else has already either done it or at least thought about it.

**Comment by OldFisher** [comment]
There is a mechanism to do something like that, but currently it works only for function wrapping:
`LUAPP_RV_CONVERT(std::vector<int>, {\
    lua::Value result = lua::Table::array(context);\
    for(auto i = 0u; i < val.size(); ++i)\
        result[i+1] = val[i];\
   return context.ret(result);\
})`

After defining that you can implicitly wrap a function that returns std::vector<int> and it will produce a table with vector's content.

I will look into extending this facility, making it usable with promotion in general. Then it will do exactly what you asked for. I'll keep this issue open until it is done.

**Comment by lambage** [comment]
Thanks, this library is amazing!!!

**Comment by lambage** [comment]
Just for the record I also needed this for function arguments and this is the macro I added

```
LUAPP_ARG_CONVERT(std::vector<uint32_t>, { \
	lua::Table table = val; \
	std::vector<uint32_t> result(table.len()); \
	auto count = 0u; \
	table.iterate([&count, &result](lua::Valref k, lua::Valref v) { \
		result[count++] = v; \
		return true; \
	}); \
	return result; \
})
```

**Comment by OldFisher** [comment]
Yes, that's how it is designed to be used. With new custom-conversion RV_CONVERT and ARG_CONVERT will have to go, but the new system will work pretty much the same.
Also, on a side note, Table::iterate iterates over every key-value pair (just like `pair` and `next` do) and in this case may produce undesirable results for tables that are not entirely arrays. Iterating numerically up to `table.len()` seems to be safer option.
